File: insertion_sort/JS/insertion.js
console.log("Isertion Sort With JS!");

// Sorts elements in ascending order by iterating through array and inserting numbers where they belong.
const insertionSort = (array)=> {
	// Iterate through each element starting at index 1.
	for(let i = 1, len = array.length; i < len; i ++) {
		let value = array[i];
		let indxB4 = i-1;
		// Compare current number to previous numbers in array.
		while(indxB4 >= 0) {
			// Insert current number before larger number.
			if(value < array[indxB4]){
				array[indxB4+1] = array[indxB4];
				array[indxB4] = value;
				indxB4--; 
			}
			else 
				break;
		}
	}
	return array;
}

console.log(insertionSort([14, 5, 0, 21, 0, 3, 4, 9, 1, 11, 10, 32, 2, 4]));

File: insertion_sort/C++/insertion.cpp
#include <iostream>

using namespace std;

int *insertion_sort(int *array, int length);

int main() {

	const int LEN = 15;
	int array[LEN] = {9,7,6,8,3,4,90,89,65,45,64,32,12,0,0};
	cout << insertion_sort(array, LEN) << endl;

	return 0;
}

int *insertion_sort(int *array, int length) {

	// Iterate through array starting from index one.
	for(int i = 1; i < length; i ++) {
		int value = array[i];
		// Index of previous number of current value.
		int indxB4 = i - 1;

		// Compare value with each element before it and keep going until no more element are greater than it.
		while(indxB4 >= 0) {
			// And place value at that position.
			if(value < array[indxB4]) {
				array[indxB4+1] = array[indxB4];
				array[indxB4] = value;
				indxB4--;
			}
			else
				break;
		}
	}
	return array;
};